securesystemslib/signer/_sigstore_signer.py

```diff
@@ -0,0 +1,156 @@
+        "keyval": {
+            "issuer": issuer,
+            "identity": identity,
+        },
+    },
+)
+
+# Create signer
+issuer = Issuer.production()
+token = issuer.identity_token()  # requires sign in with GitHub in a browser
+signer = SigstoreSigner(token, public_key)
+
+# Sign
+signature = signer.sign(b"data")
+
+# Verify
+public_key.verify_signature(signature, b"data")
+
+```
+
+"""
+
+import io
+import logging
+from typing import Any, Dict, Optional
+
+from securesystemslib import exceptions
+from securesystemslib.signer._signer import (
+    Key,
+    SecretsHandler,
+    Signature,
+    Signer,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class SigstoreKey(Key):
+    """Sigstore verifier."""
+
+    @classmethod
+    def from_dict(cls, keyid: str, key_dict: Dict[str, Any]) -> "SigstoreKey":
+        keytype = key_dict.pop("keytype")
+        scheme = key_dict.pop("scheme")
+        keyval = key_dict.pop("keyval")
+
+        for content in ["identity", "issuer"]:
+            if content not in keyval or not isinstance(keyval[content], str):
+                raise ValueError(
+                    f"{content} string required for scheme {scheme}"
+                )
+
+        return cls(keyid, keytype, scheme, keyval, key_dict)
+
+    def to_dict(self) -> Dict:
+        return {
+            "keytype": self.keytype,
+            "scheme": self.scheme,
+            "keyval": self.keyval,
+            **self.unrecognized_fields,
+        }
+
+    def verify_signature(self, signature: Signature, data: bytes) -> None:
+        from sigstore.verify import VerificationMaterials, Verifier
+        from sigstore.verify.policy import Identity
+        from sigstore_protobuf_specs.dev.sigstore.bundle.v1 import Bundle
+
+        verifier = Verifier.production()
+        identity = Identity(
+            identity=self.keyval["identity"], issuer=self.keyval["issuer"]
+        )
+
+        signature_bundle = Bundle().from_dict(signature.signature)
+        verification_materials = VerificationMaterials.from_bundle(
+            input_=io.BytesIO(data), bundle=signature_bundle, offline=True
+        )
+
+        try:
+            result = verifier.verify(verification_materials, identity)
+            if not result:
+                logger.info(
+                    "Key %s failed to verify sig: %s", self.keyid, result.reason
+                )
+                raise exceptions.UnverifiedSignatureError(
+                    f"Failed to verify signature by {self.keyid}"
+                )
+        except exceptions.UnverifiedSignatureError:
+            raise
+
+        except Exception as e:
+            logger.info("Key %s failed to verify sig: %s", self.keyid, str(e))
+            raise exceptions.VerificationError(
+                f"Unknown failure to verify signature by {self.keyid}"
+            ) from e
+
+
+class SigstoreSigner(Signer):
+    """Sigstore signer."""
+
+    def __init__(self, token: str, public_key: Key):
```

Comment:
Requiring the token to be created beforehand is not very nice UX. The interface here should look more like the sigstore sign cli, where the user can pass identity/issuer, or instruct it to use ambient credentials.

Reply:
Yeah makes sense -- I guess the most important bit is how it works with from_private_key_uri(). The ways to deliver the "arguments" are then
* signing key specific details included in private_key_uri
* signing system variables can also be "ambient" (so environment variables etc)

Reply:
You are right, the preferred signer creation UI is `from_private_key_uri`. So creating the token there and passing it to the constructor should work.